keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,974 @@
+add it again because it is redundant. It makes log output larger (for
+performance reasons, loggers like zap and klog do not filter out duplicates)
+and causes additional overhead.
+
+The guiding principle is that the callee should only add key/value pairs for
+information that the caller doesn’t have, like a local variable, and leave the
+decision about adding information about the parameters to the caller.
+
+Exceptions are when a log message is related to a parameter or a part of it
+that normally doesn’t need to be logged.
+
+This needs to be checked by taking a holistic view and ideally documenting the
+logging behavior.
+
+This situation is similar to wrapping errors: whether an error message contains
+information about a parameter (for example, a path name) needs to be documented
+to avoid storing redundant information when the caller wraps an error that was
+created by the callee.
+
+Analysis of the JSON log files collected at a high log level during a Prow test
+run can be used to detect cases where redundant key/value pairs occur in
+practice. This is not going to be complete, but it doesn’t have to be because
+the additional overhead for redundant key/value pairs matters a lot less when
+the message gets emitted infrequently.
+
+#### Modifying the same variable in a loop
+
+Reusing variable names keeps the code readable and prevents errors because the
+variable that isn’t meant to be used will be shadowed. However, care must be
+taken to really create new variables. This is broken:
+
+```go
+func foo(logger klogr.Logger, objects ...string) {
+    for _, obj := range objects {
+        // BUG: logger accumulates different key/value pairs with the same key
+        logger = logger.WithValue(“obj”, obj)
+        doSomething(logger, obj)
+    }
+}
+```
+
+A new variable must be created with `:=` inside the loop:
+
+```go
+func foo(logger klogr.Logger, objects ...string) {
+    for _, obj := range objects {
+        // This logger variable shadows the function parameter.
+        logger := logger.WithValue(“obj”, obj)
+        doSomething(logger, obj)
+    }
+}
+```
+
+#### Unused return values
+
+This code looks like it adds a name, but isn’t doing it correctly:
+
+```go
+func foo(logger klogr.Logger, objects ...string) {
+        // BUG: WithName returns a new logger with the name,
+        // but that return value is not used.
+        logger.WithName(“foo”)
+        doSomething(logger)
+    }
+}
+```
+
+
+## Design Details
+
+klog currently provides a formatter for log messages, a global logger instance,
+and some helper code. It also contains code for [log
+sanitization](https://github.com/kubernetes/enhancements/issues/1753), but that
+is a [deprecated](https://github.com/kubernetes/enhancements/pull/3096) alpha
+feature and doesn't need to be supported anymore.
+
+A new package `k8s.io/klogr` will replace both `k8s.io/klog` and
+`k8s.io/klog/klogr`. It will be hosted under `staging` and thus get developed
```

Comment:
Currently, k8s.io/kubernetes has many dependencies which use k8s.io/klog. We would not be able to pull in updated versions of those dependencies that switched to use k8s.io/klogr if k8s.io/klogr is a staging repo.

Reply:
Let me try to understand. Suppose k/k depends on example.com/package which depends on klog. There's then a vendor/example.com/package directory with the vendored code. Why is that not possible when the dependency is k/k -> example.com/package -> k8s.io/klogr -> klog?

Ah, I think I understand: we cannot break the API of k8s.io/klogr because that breaks example.com/package, but example.com/package cannot be updated because we cannot merge the k8s.io/klogr API change.

Let me revise the KEP proposal then: we need a stand-alone repo similar to klog. It won't have a stable API initially, but that's okay. The code in it is fairly minimal, we should be able to reach v1 soon.

Reply:
looks like this thread is still unresolved... KEP still indicates this will be hosted under `staging`

It's also not clear to me what the proposed dependency graph looks like. Does k8s.io/klogr depend on k8s.io/klog, or vice-versa (or both)?

Reply:
I missed the "staging" part under "design details" and only updated the "Infrastructure Needed" section.

It needs to be a stand-alone repo just like `k8s.io/klogr`. Initially it will have `k8s.io/klog` as dependency and other than that just the standard Go runtime. In the final phase (GA) we drop that `klog.io/klog` dependency.

Reply:
I'm not sure I see the problem.

if k/k depends on foo and foo depends on klogr, and we vendor foo into k/k, klogr will resolve to ./staging/.../klogr - right?

Reply:
`k8s.io/klog` will not depend on `k8s.io/klogr`.

Reply:
still a distinct repo seems better to me, unless we think this will actually get a lot of churn (and we still have to maintain a sane API)